### ISSUE TITLE
Twisted adapter reconnection on connection lost

### ISSUE THREAD

**Issue by nanvel** [pika/pika#820]
This is a code bellow I am able to reproduce the issue: run it and restart rabbitmq, first it hangs and in a few minutes I see errors in logs, it doesn't reconnect.

I was googling and digging into the code and still can't make it works (reconnect).
The last idea was to make `protocol._InstanceFactory` to support reconnection, same `protocol.ReconnectingClientFactory`. Another idea was using `on_connection_closed`, just call `connect` method to reconnect, but then I need to write some logic to unblock and resend messages were hanged, and it looks not reliable.

I created an issue for it because:
- it is an extremely important feature for a system reliability, and there must be an easy way to turn it on
- and if the way exists, it must be documented

I am looking for examples or suggestions for the right way to implement reconnection. 

```python
# pika==0.10.0
# Twisted==17.1.0
import logging

import pika
from pika.adapters import twisted_connection
from twisted.internet import defer, protocol, reactor


logging.basicConfig(level=logging.DEBUG)


class MQ:

    EXCHANGE_NAME = 'my-exchange'

    def __init__(self, rabbitmq_host, rabbitmq_port, rabbitmq_user, rabbitmq_pass, rabbitmq_vhost):
        self.rabbitmq_host = rabbitmq_host
        self.rabbitmq_port = rabbitmq_port
        self.rabbitmq_user = rabbitmq_user
        self.rabbitmq_pass = rabbitmq_pass
        self.rabbitmq_vhost = rabbitmq_vhost

        self.connection = None
        self.channel = None

        self._closing = False
        self._sent = 0
        self._deliveries = {}

    @defer.inlineCallbacks
    def connect(self):
        if self.rabbitmq_user:
            credentials = pika.PlainCredentials(
                username=self.rabbitmq_user,
                password=self.rabbitmq_pass
            )
        else:
            credentials = None
        parameters = pika.ConnectionParameters(
            credentials=credentials,
            virtual_host=self.rabbitmq_vhost
        )
        cc = protocol.ClientCreator(
            reactor,
            twisted_connection.TwistedProtocolConnection,
            parameters
        )
        df = cc.connectTCP(host=self.rabbitmq_host, port=int(self.rabbitmq_port))
        df.addCallback(lambda p: p.ready)
        self.connection = yield df

        self.connection.add_on_close_callback(self.on_connection_closed)

        self.channel = yield self.connection.channel()
        self.channel.confirm_delivery(callback=self.on_delivery_confirmation)
        yield self.channel.exchange_declare(
            exchange=self.EXCHANGE_NAME,
            exchange_type='topic',
            durable=True,
            auto_delete=False
        )

    def on_connection_closed(self, connection, reply_code, reply_text):
        if not self._closing:
            print("Connection failure ...")

    def on_delivery_confirmation(self, method_frame):
        """
        http://pika.readthedocs.io/en/0.10.0/examples/asynchronous_publisher_example.html
        """
        delivery_tag = method_frame.method.delivery_tag
        for key in list(self._deliveries.keys()):
            if key <= delivery_tag:
                self._deliveries[key].callback(None)
                del self._deliveries[key]
                del self._messages[key]

    @defer.inlineCallbacks
    def close(self):
        self._closing = True
        yield self.connection.close()

    @defer.inlineCallbacks
    def send(self, message):
        """
        Send and async wait for delivery.
        """
        delivery = defer.Deferred()
        self._sent += 1
        self._deliveries[self._sent] = delivery
        yield self.channel.basic_publish(
            exchange=self.EXCHANGE_NAME,
            routing_key='my-routing-key',
            body=message,
            properties=pika.BasicProperties(content_type='text/plain', content_encoding='utf8')
        )
        yield delivery


if __name__ == '__main__':
    mq = MQ(
        rabbitmq_host='localhost',
        rabbitmq_vhost=None,
        rabbitmq_port=5672,
        rabbitmq_user='',
        rabbitmq_pass=''
    )

    @defer.inlineCallbacks
    def run():
        yield mq.connect()
        for i in range(5000):
            yield mq.send(f"Hi {i}!")
            print(f"{i} sent.")
        yield mq.close()
        reactor.stop()

    reactor.callWhenRunning(run)
    reactor.run()
```

**Comment by lukebakken** [comment]
There is what appears to be a pretty complete code example [here](https://github.com/pika/pika/blob/master/examples/twisted_service.py). If that doesn't demonstrate what you are looking for, feel free to re-open this issue. Thanks!
